ls1intum/Artemis
# `Programming exercises`: Generate solution entries for behavioral test cases

### Checklist
#### General


#### Server
#### Client
#### Changes affecting Programming Exercises
- [ ] ~~I tested **all** changes and their related features with **all** corresponding user types on Test Server 2 (Jenkins and Gitlab).~~ Testwise coverage not supported here

### Motivation and Context


These changes are part of the Bachelor Thesis from @Hialus and @ole-ve about the automatic generation of code hints for programming exercises (HESTIA) in Java.
A previous PR already introduced the ability to create solution entries for structural test cases. This PR tackles solution entries for behavioral test cases using the git-diff and testwise coverage reports.
A reminder: Solution entries are a single code block that the students have to change in the template to get closer to passing a test case.

### Description

The process of generating solution entries for behavioral test cases is a bit complicated, but I'll try to give a brief rundown below. Due to this and the fact that the process should be easily extendable in the future, I used a slightly modified Blackboard pattern for this. This means that all steps are split up into knowledge sources and new knowledge sources can easily be added in the future.

Requirements:
- The git-diff report and testwise coverage report must exist
- The solution build plan must have run through without a build failure

Steps:
1. `GroupGitDiffAndCoverageEntriesByFilePathAndTestCase`: This will create GroupedFiles that bundle all ProgrammingExerciseGitDiffEntries and TestwiseCoverageReportEntries of a single file and test case together
2. `ExtractCoveredLines`: A small step that extracts all lines that were covered by the TestwiseCoverageReportEntries
3. `ExtractChangedLines`: A small step that extracts all lines that were changed in the ProgrammingExerciseGitDiffEntries
4. `FindCommonLines`: The intersection of the previous two lists
5. `CreateCommonChangeBlocks`: Takes the output of the previous step and creates GroupedFile.ChangeBlock instances from it. Each code block represents a single string of continuous common lines
6. `InsertFileContents`: Inserts the contents of each file in the solution repository into their related GroupedFiles
7. `AddUncoveredLinesAsPotentialCodeBlocks`: There are certain lines that are not covered by Jacoco, but are still relevant for the student or can improve the readability of the generated solution entries. These lines include else statements, curly brackets, and empty lines. 
This step will add those lines as potential code blocks, which will only be added in certain circumstances (This will be part of a future PR. Currently all code blocks are included in the end result)
8. `CombineChangeBlocks`: The previous step can lead to directly adjacent or intersecting code blocks within a single grouped file. This step combines those code blocks into one.
9. `CreateSolutionEntries`: Creates a solution entry for every code block in every grouped file

### Steps for Testing

As there still is not proper view for solution entries in the client you will again need to rely on the developer console output. #4976 Will include the first part of this, so if that PR gets merged you can use that.

Prerequisites:
- 1 Admin

1. Log in to Artemis
2. Create a Gradle_Gradle Java Programming exercise with Testwise Coverage option enabled
3. Do any changes to the test and solution repo that you want
4. Wait for the build of the solution build plan to finish
5. Open the developer console
6. Press the "Create behavioral solution entries button"
7. Check that the printed solution entries are correct (Cross-reference with the git-diff and coverage reports and under consideration of the AddUncoveredLinesAsPotentialCodeBlocks step)

### Review Progress





#### Code Review
#### Manual Tests

### Test Coverage


| Class/File | Branch | Line |
|------------|-------:|-----:|
| All Knowledge sources | >80% | >77% |
| BehavioralBlackboard | n/a | 100% |
| BehavioralSolutionEntryGenerationException | n/a | 50% |
| BehavioralTestCaseService | 73% | 88% |
| GroupedFile | 58% | 98% |
| GroupedFile.ChangeBlock | 61% | 72% |
